pyfr/integrators/multip.py

```diff
@@ -0,0 +1,150 @@
+# -*- coding: utf-8 -*-
+
+import re
+
+from pyfr.inifile import Inifile
+from pyfr.integrators.dual.base import BaseDualIntegrator
+from pyfr.util import proxylist
+
+
+class MultiP(BaseDualIntegrator):
+    def __init__(self, *args, **kwargs):
+        # Get config file and order
+        cfg = kwargs['cfg']
+        order = cfg.getint('solver', 'order')
+
+        # Get multigrid cycle
+        mgsect = 'pseudo-multigrid'
+        self.lvliters = list(cfg.getliteral(mgsect, 'pseudo-mgcycle'))
+        self.nlevels = len(self.lvliters)
+        self.level = 0
+
+        if self.nlevels > order:
+            raise ValueError('The number of multigrid levels cannot exceed '
+                             'the solution order')
+
+        # The maximum and minimum number of multigrid cycles
+        self.maxcycles = cfg.getint(mgsect, 'pseudo-mgcycles-max')
+        self.mincycles = cfg.getint(mgsect, 'pseudo-mgcycles-min')
+
+        if self.maxcycles < self.mincycles:
+            raise ValueError('The maximum number of multigrid cycles must '
+                             'be greater than or equal to the minimum')
+
+        # Multigrid pseudo-time steps
+        dtau = cfg.getfloat('solver-time-integrator', 'pseudo-dt')
+        dtaufact = cfg.getfloat(mgsect, 'pseudo-dt-fact', 1.0)
+        self.dtaus = [dtau*dtaufact**i for i in range(self.nlevels)]
+
+        # Generate multiple cfgs for the multigrid systems
+        self.mgcfg = [Inifile(cfg.tostr()) for i in range(self.nlevels)]
+        for deg in range(1, order + 1):
+            self.mgcfg[deg - 1].set('system', 'order', order - (deg - 1))
+            for sec in cfg.sections():
+                m = re.match(r'solver-(.*)-mg-p{0}'.format(deg), sec)
+                if m:
+                    self.mgcfg[order - deg].rename_section(
+                        m.group(0), 'solver-' + m.group(1)
+                        )
+
+        super().__init__(*args, **kwargs)
+
+    # Patch idxcurr
+    @property
+    def _idxcurr(self):
+        return self._mgidxcurr[self.level]
+
+    @_idxcurr.setter
+    def _idxcurr(self, y):
+        self._mgidxcurr[self.level] = y
+
+    # Patch regs
+    @property
+    def _regs(self):
+        return self._mgregs[self.level]
+
+    @_regs.setter
+    def _regs(self, y):
+        self._mgidxcurr[self.level] = y
+
+    # Patch system
+    @property
+    def system(self):
+        return self._mgsystem[self.level]
+
+    def _init_reg_banks(self, nreg):
+        self._mgregs, tmp, self._regidx = [], [], list(range(nreg))
+        self._mgidxcurr = [0]*self.nlevels
+
+        # Create a proxylist of matrix-banks for each storage register
+        for l in range(self.nlevels):
+            for i in self._regidx:
+                tmp.append(
+                    proxylist([self.backend.matrix_bank(em, i)
+                               for em in self._mgsystem[l].ele_banks])
+                )
+            self._mgregs.append(tmp)
```

Comment:
Could you elaborate? Now we have regidx=[0,1,2] and they point to registers 
[[pl(r0),pl(r1),pl(r2)],[pl(r0),pl(r1),pl(r2)],[pl(r0),pl(r1),pl(r2)]] 
where each entry is a proxylist taking care of multiple element types. Regidx does not necessarily have to be level dependent since the registers are always in same phase between levels. We just need to track idxcurr that is level dependent